src/Psl/Xml/Internal/stop_on_first_issue.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Xml\Internal;
+
+use Generator;
+use Psl\Xml\Exception\RuntimeException;
+
+use function libxml_clear_errors;
+use function libxml_get_errors;
+use function libxml_use_internal_errors;
+
+/**
+ * Additional logic needs to be placed on top of the detect_errors function if you want to lazily detect errors.
+ * This method can be used while reading an \XMLReader and returning a Generator.
+ * After every tick, this function will validate if issues were detected.
+ *
+ * @internal
+ *
+ * @template Tr
+ *
+ * @psalm-param callable(): bool $tick
+ * @psalm-param callable(): Tr|null $read
+ * @psalm-return Generator<int, Tr, mixed, null>
+ *
+ * @throws RuntimeException
+ */
+function stop_on_first_issue(callable $tick, callable $run): Generator
```

Comment:
Example usage:

```php

<?php
$reader = new XMLReader();
$reader->open('users.xml');

$iterator = stop_on_first_issue(
    static fn(): bool => $reader->read(),
    static fn(): ?string => $reader->name === 'user' ? $reader->readOuterXml() : null,
);

// This will not report any issues and trigger warnings, because of the Generator
// If you replace the yield with e.g. an array, it will work
foreach ($iterator as $xml) {
    echo $xml;
}
```